### ISSUE TITLE
Eng Design: A ranking algorithm for revisions to be reviewed

### ISSUE THREAD

**Issue by xinbenlv** [google/wikiloop-battlefield#112]
### How do we identify the ***qualified factors of priority***? 

In general, we want about 500 - 2000 revisions, denoted as `REVIEW_QUEUE_SIZE`, and then we use the following a few aspects for a quick priorities:

1. **freshness**, the fresher a revision, the more often we want a reviewer
   to work on.

2. **whether it's a hard case**: if a revision has hard case, we want
   more pairs of eyes to review them them, such as:
   - when there are a number of judgement of *Unsure*
   - when the AI vandalism detection scores,such as ORES, ClueBotNG and STiki
    conflict with each other, or
   - when future *active learning* algorithm suggest prioritizing certain cases, if labelled, will maximize the improvement of a machine learning model.

3. **editor-selected importance**: whether the page is considered higher priority
   by the individual editor. For example, if the page happen to be on the watch-list
   of the editor, or falls into some categories that editors collaboratively
   contribute to. To make it super simple with a single entry point for a user,
   we include this feature here, so we can store a persistent configuration of a
   users' importance factor.
   instead of letting user to choose specific parameter each time each time.

4. **random sample**: we will also add a few randomly acquired revisions.

### How do we ***quantify*** the qualified factors for priorities? 

In order to quantify the factors, so that we can represent the priority with a
numeric score and use such priority score to rank the score, we need to decide
a quantity unit.

The most significant existing numeric measure is the freshness of a revision.
Two options we have:
  1. use the `Timestamp` with a unit of second,
  2. use the `RevisionId`, with a unit of individual revision.

The similarity between them is that both monotonically increase with time, the newer a revision
is, the larger these numbers are.

The difference are, the distributions of these numbers.

If we use `Timestamp`, a timestamp increase linearly with global absolute time. It can be compared  across different languages of Wikis. So if we later decide to support multiple Wikis in a single review queue, we can do it without much additional difficulty - all other factors are already represent in a globally-normalized metrics.

If we use `RevisionId`, such id number increase linearly with the number of revisions. Therefore, if we factor other priority-related quantity in this unit, the relative ranking will remain similar across Wikis that has significant different rate of new edits.

### Other considerations
Other than the qualified and quantified approach, we will also include the following considerations for a solution:

1. How extendable it is for other potential new priorities introduced by WikiLoop Battlefield developers or community members. 
2
. The User Experience of an reviewer of WikiLoop Battlefield: 
 - will it be boring?
 - will it make a good use of the reviewer's precious time?
 - will it encourage collaboration between different reviewers?
 - will it encourage taking action such as thanking an edit, or reverting an edit?
3. Performance of `client` and `server`
 - whether we can batch supply the information, and how complex to implement?
 - whether we can query, cache-store and implement on client side, and how complex it will be?

**Comment by xinbenlv** [comment]
Using npm.com/heap, but blocked by https://github.com/vuejs/vuex/issues/1437: in vuex, a state mutation and commit is separate, so 
    `hap.pop(); `
is not available.

**Comment by xinbenlv** [comment]
@hrasyid Here you go. And the code location: https://github.com/google/wikiloop-battlefield/blob/41105a196cfa4406e4da0e94ea270e328ea4615f/store/revisions.js#L16

**Comment by hrasyid** [comment]
Agree with most of your points above.
Some additional points that I think we should consider

- Somewhat bump priority of edits from non signed-in editors (higher risk usually)
- Lower priority for edits marked as bots (bots are usually vetted before they can use the "bot" marker, therefore lower priority)
- Allow reviewers to have a list of "Trusted" editors (existing people that they know wouldn't vandalize), so they can be excluded from the review queue.
